app/javascript/hooks/useLazyFetch.ts

```diff
@@ -0,0 +1,190 @@
+import React from "react";
+
+import { assertResponseError, request } from "$app/utils/request";
+
+import { showAlert } from "$app/components/server-components/Alert";
+
+interface UseLazyFetchOptions<T> {
+  url: string;
+  responseParser: (data: unknown) => T;
+  hasMore?: boolean;
+}
+
+interface UseLazyFetchResult<T> {
+  data: T;
+  isLoading: boolean;
+  setIsLoading: (isLoading: boolean) => void;
+  setData: (data: T) => void;
+  fetchData: (queryParams?: QueryParams) => Promise<void>;
+  hasLoaded: boolean;
+  setHasLoaded: (hasLoaded: boolean) => void;
+}
+
+type QueryParams = Record<string, string | number>;
+
+export type Pagination = {
+  count: number;
+  from: number;
+  in: number;
+  last: number;
+  limit: number;
+  next: number | null;
+  offset: number;
+  outset: number;
+  overflow: number;
+  page: number;
+  prev: number | null;
+  to: number;
+};
+
+type PaginatedResponse = {
+  pagination: Pagination;
+};
+
+const isPaginatedResponse = (data: unknown): data is PaginatedResponse => {
+  if (typeof data !== "object" || data === null || !("pagination" in data)) {
+    return false;
+  }
+  return typeof data.pagination === "object" && data.pagination !== null;
+};
+
+// Internal hook that handles the core fetching logic
+const useLazyFetchCore = <T>(
+  initialData: T,
+  options: UseLazyFetchOptions<T>,
+  shouldFetchCondition: (hasLoaded: boolean) => boolean,
+  onSuccess?: (responseData: unknown, parsedData: T) => void,
+) => {
+  const [data, setData] = React.useState<T>(initialData);
+  const [isLoading, setIsLoading] = React.useState(true);
+  const [hasLoaded, setHasLoaded] = React.useState(false);
+
+  const fetchData = React.useCallback(
+    async (queryParams: QueryParams = {}) => {
+      if (!shouldFetchCondition(hasLoaded)) return;
```

Comment:
I find it a bit odd that the data won't necessarily be fetched if we explicitly call `fetchData` (and the condition in the hook parameters might be a long way from the `fetchData` call)

Would it perhaps work to refactor this to have an `isEnabled` parameter to control whether to fetch at all (i.e. should be set to the value of `isOpen` in all the expandable sections). We could optionally return a `fetchData` function as well which always forces a fetch

Note: This is roughly inspired by Tanstack Query, which we found great to use in Helper

I also feel like we could put the `page` param logic into `useLazyPaginatedFetch` since we expect a standard pagination object there anyway. So to supplement `isEnabled` for that hook we could return a `fetchNextPage` function

Reply:
I've refactored it so `fetchData` always fetches. Took a little different approach for the test, with a `fetchOnMount` option that is more descriptive than `isEnabled` IMO. Let me know what you think